### PR TITLE
Rearrange order of SIM_ARGS, EXTRA_ARGS to match VVP restrictions

### PR DESCRIPTION
**Pull request by mballance**
VVP requires simulator options and arguments (those that control behavior of the VVP simulator) to be passed before the .vvp file. Based on the documentation, SIM_ARGS are:
```
Any arguments or flags to pass to the execution of the compiled simulation.  Only applies to simulators with a separate compilation stage (currently Icarus and VCS).
```

In the current order, simulator options (eg -M/-m to load VPI libraries) cannot be controlled via the cocotb Makefiles. 

The change in this PR, as discussed with @themperek in #1443, is to pass `$(SIM_ARGS)` and `$(EXTRA_ARGS)` before the .vvp file is specified.

Signed-off-by: Matthew Ballance <matt.ballance@gmail.com>

### CODE REVIEW

**Comment by ktbarrett** [comment]
Does this change the contents of `cocotb.argv`?

**Comment by mballance** [comment]
> Does this change the contents of `cocotb.argv`?

The order of arguments will change, but all the arguments that were present before will still be there with this change. 

`cocotb.argv` is set based on the content of  `argv` returned by `vpi_get_vlog_info`.  The arguments returned by `vpi_get_vlog_info` are the entire simulator command line, including tool-specific options.

**Comment by themperek** [comment]
Some "interesting" error here: https://travis-ci.org/cocotb/cocotb/jobs/654110146#L2112-L2133

**Comment by themperek** [comment]
@eric-wieser CI error related to #1440 ?

**Comment by mballance** [comment]
> @eric-wieser CI error related to #1440 ?

Several of the tests (I'm looking at test_plusargs) look like they're trying to add -v93 to SIM_ARGS if the simulator is NCSIM (ius, xcelium):

```
ifeq ($(SIM),$(filter $(SIM),ius xcelium))
    SIM_ARGS += -v93
endif
```

The core ius and Makefiles also adds this to SIM_ARGS:
```
GPI_EXTRA:=
ifeq ($(TOPLEVEL_LANG),verilog)
    GPI_ARGS = -loadvpi $(LIB_DIR)/libvpi
    EXTRA_ARGS += -v93
    HDL_SOURCES = $(VERILOG_SOURCES)
    ROOT_LEVEL = $(TOPLEVEL)
ifneq ($(VHDL_SOURCES),)
    HDL_SOURCES += $(VHDL_SOURCES)
    GPI_EXTRA = cocotbvhpi
endif
else ifeq ($(TOPLEVEL_LANG),vhdl)
    GPI_EXTRA = cocotbvhpi
    EXTRA_ARGS += -v93
    EXTRA_ARGS += -top $(TOPLEVEL)
    RTL_LIBRARY ?= $(TOPLEVEL)
    MAKE_LIB = -makelib $(RTL_LIBRARY)
    HDL_SOURCES = $(VHDL_SOURCES)
ifneq ($(VERILOG_SOURCES),)
    HDL_SOURCES += $(VERILOG_SOURCES)
endif
else
   $(error "A valid value (verilog or vhdl) was not provided for TOPLEVEL_LANG=$(TOPLEVEL_LANG)")
endif
```

I see two options:
- debug the conditional in the test-specific Makefiles to understand why -v93 is being passed to Icarus
- Rely on the IUS/Xcellium Makefiles to add -v93 to SIM_ARGS

**Comment by ktbarrett** [comment]
Not specifying `SIM` used to default to `icarus`. Now it doesn't, and that conditional evaluates to true because `""` == `""`.

**Comment by cmarqu** [comment]
> The core ius Makefile also adds this to SIM_ARGS:

> * Rely on the IUS/Xcellium Makefiles to add -v93 to SIM_ARGS

When I introduced the ``Makefile.xcelium``, I removed the ``-v93`` from the default ``SIM_ARGS`` because this would preclude having other conflicting options used in user Makefiles. ``Makefile,ius`` is gradually becoming obsolete because IUS/Incisive stopped being developed by Cadence in 2015, so I left the options untouched there so that there would not be a backward compatibility problem.
In short, adding ``-v93`` by default would be a step backward.

**Comment by mballance** [comment]
> > The core ius Makefile also adds this to SIM_ARGS:
> 
> > * Rely on the IUS/Xcellium Makefiles to add -v93 to SIM_ARGS
> 
> When I introduced the `Makefile.xcelium`, I removed the `-v93` from the default `SIM_ARGS` because this would preclude having other conflicting options used in user Makefiles. `Makefile,ius` is gradually becoming obsolete because IUS/Incisive stopped being developed by Cadence in 2015, so I left the options untouched there so that there would not be a backward compatibility problem.
> In short, adding `-v93` by default would be a step backward.

Thanks for the input, @cmarqu. How about if change the conditional in the test Makefiles? We could have separate checks for ius and xcelium. In other words:

```
ifeq ($(SIM),"ius")
SIM_ARGS += -v93
endif
ifeq ($(SIM),"xcelium")
SIM_ARGS += -v93
endif
```
Not a creative solution, but should be functional in the presence of SIM not being set. Thoughts or other suggestions?

**Comment by cmarqu** [comment]
> How about if change the conditional in the test Makefiles? We could have separate checks for ius and xcelium.

In that case, we could even leave ``ius`` out since the core Makefile already sets the ``-93``(we need to remember not to remove it), but maybe we could have an extra check for "is ``SIM`` empty or just whitespace" with 
```
ifeq ($(strip $(SIM)),)
# SIM not given
else ifeq ($(SIM),$(filter $(SIM),ius xcelium))
    SIM_ARGS += -v93
endif
```
(untested). Ah, I now read that the ``else ifeq`` is not supported in make 3.80, but maybe we don't need to care for this old version? This make stuff is really tiring...

**Comment by mballance** [comment]
@cmarqu, good point on just adding a layer of nesting to avoid getting tripped up by an empty `$(SIM)`. I'll update the PR with this change

**Comment by themperek** [comment]
Could you split this to 2 diffrent PRs. It looks not really related to me.

File: tests/designs/vhdl_configurations/Makefile [hunk 1]
```diff
@@ -34,8 +34,10 @@ else
 endif
 VHDL_SOURCES += $(COCOTB)/tests/designs/vhdl_configurations/configurations.vhd
 
-ifeq ($(SIM),$(filter $(SIM),ius xcelium))
-    SIM_ARGS += -v93
+ifneq (,$(SIM))
```
Comment (eric-wieser) [inline tests/designs/vhdl_configurations/Makefile]: I don't get this - is `SIM` ever blank?

File: tests/designs/viterbi_decoder_axi4s/Makefile [hunk 2]
```diff
@@ -65,8 +65,10 @@ VHDL_SOURCES = $(SRC_BASE)/packages/pkg_helper.vhd \
 	       $(SRC_BASE)/src/recursion.vhd \
 	       $(SRC_BASE)/src/dec_viterbi.vhd
 
-ifeq ($(SIM),$(filter $(SIM),ius xcelium))
-    SIM_ARGS += -v93
+ifneq (,$(SIM))
+    ifeq ($(SIM),$(filter $(SIM),ius xcelium))
+        SIM_ARGS += -v93
+    endif
 endif
```
Comment (eric-wieser) [inline tests/designs/viterbi_decoder_axi4s/Makefile]: Simpler as:
```
ifneq ($(filter $(SIM),ius xcelium),)
    SIM_ARGS += -v93
endif
```

**Comment by mballance** [comment]
> Could you split this to 2 different PRs? They look not really related to me.

I considered this fix related, since it was required in order for #1447 to pass tests. If splitting into two is the shortest path to getting both in, I can certainly do that. Please confirm.

File: tests/designs/vhdl_configurations/Makefile (hunk 1 above)
Comment (mballance) [reply tests/designs/vhdl_configurations/Makefile]: According to @ktbarrett, SIM is now blank if not specified by the user. It used to be defaulted to "icarus". Now, blank is equivalent to "icarus".

File: tests/designs/viterbi_decoder_axi4s/Makefile (hunk 2 above)
Comment (mballance) [reply tests/designs/viterbi_decoder_axi4s/Makefile]: That's fair, @eric-wieser. Let's see where we're actually putting this fix, and I'll make the changes.

**Comment by themperek** [comment]
Yes. I think it is better to split.

SIM is not empty https://github.com/cocotb/cocotb/blob/master/cocotb/share/makefiles/Makefile.inc#L68

File: tests/designs/viterbi_decoder_axi4s/Makefile (hunk 2 above)
Comment (mballance) [reply tests/designs/viterbi_decoder_axi4s/Makefile]: Focused this PR on Icarus argument passing

**Comment by themperek** [comment]
@mballance Why is CI failing now?

**Comment by mballance** [comment]
@themperek, it's because somehow the if conditionals are triggering and adding `-v93` to SIM_ARGS when running Icarus. Before #1447, this wouldn't trigger a failure because SIM_ARGS were to the right of the .vvp file. Icarus considers any arguments to the right of the .vvp file to be uninterpreted. Now that SIM_ARGS are on the left, Icarus is complaining that it doesn't accept `-v93`... Here's the output from Travis:

```
  TESTCASE= TOPLEVEL=tb_top TOPLEVEL_LANG=verilog COCOTB_SIM=1 \
        /usr/local/bin/vvp -M /Users/travis/build/cocotb/cocotb/.tox/py37/lib/python3.7/site-packages/cocotb/libs/icarus -m gpivpi -v93  sim_build/sim.vvp +foo=bar +test1 +test2 +options=fubar
/usr/local/bin/vvp: illegal option -- 9
/usr/local/bin/vvp: illegal option -- 3
make[4]: *** [results.xml] Error 2
make[3]: *** [sim] Error 2
make[3]: Target `all' not remade because of errors.
make[2]: *** [test_plusargs] Error 2
```

**Comment by themperek** [comment]
OK. So let's try to fix one at a time. Sorry for all this. If you could make be PR to fix this first.

**Comment by ktbarrett** [comment]
@themperek I've noticed this when I was working on my custom entry point PR that those default assignments in Makefile.inc aren't working. Not sure why. I added `GPI_EXTRA ?= cocotb` directly above `SIM ?= icarus` and it didn't take. The only reason Matt is seeing those issues is because it's not working. So that needs some thought, but that's a separate issue. I think the current makefiles are masking an actual issue.

**Comment by themperek** [comment]
@ktbarrett No I think  I know what happened with #1306 I had to move `SIM ?= icarus` to `Makefile.inc`. In general, I would like to remove `Makefile.inc` and move everything to `Makefile.sim` but probably in a few steps. Maybe should make a PR and move https://github.com/cocotb/cocotb/blob/master/cocotb/share/makefiles/Makefile.inc#L67-L88 to `Makefile.sim` first.

**Comment by themperek** [comment]
Maybe we can replace with this?
```Makefile
ifneq (,$(findstring $(SIM),ius xcelium))
    SIM_ARGS += -v93
endif
```
@cmarqu

**Comment by eric-wieser** [comment]
That's what I suggest in https://github.com/cocotb/cocotb/pull/1447#discussion_r383220980, so I'm on board with that.

**Comment by ktbarrett** [comment]
Something's up with the Travis reporter, the tests all pass.
